### PR TITLE
Full forward pass with argmax, final proj and causal mask

### PR DESCRIPTION
**Pull request by nikkolasg**

### CODE REVIEW

File: zkml/src/layers/transformer/softmax.rs [hunk 1]
```diff
@@ -7,15 +9,29 @@ use crate::{
 
 #[derive(Debug, Clone)]
 pub struct Softmax<N> {
+    // By default, it's equal to 1
     pub scale: N,
+    // By default, softmax is going to be applied on the full tensor.
+    // You can specificy a dimen to apply softmax on. For example, for a tensor  of shape [2,3,4],
+    // if apply_on_dim = 2, then softmax will be applied on every chunks of 3 * 4 elements each.
```
Comment (nicholas-mainardi) [inline zkml/src/layers/transformer/softmax.rs:16]: Shoudln't this be `apply_on_dim = 0`?

File: zkml/src/layers/transformer/mod.rs [hunk 2]
```diff
@@ -174,61 +221,86 @@ mod test {
...
             if let Some(gpt2_output) = gpt2_output {
-                assert!(gpt2_output.is_qkv_close(qkv.outputs()));
+                ensure!(gpt2_output.is_qkv_close(qkv.outputs()));
             }
             let mha = self
                 .mha
                 .evaluate::<GoldilocksExt2>(&qkv.outputs(), vec![])?;
+            // first, we need to apply causal mask before softmax
+            // let causal_mask = causal_mask(self.num_heads, seq_len, seq_len);
+            // let masked_attention_scores = mha.outputs()[0].add(&causal_mask);
```
Comment (nicholas-mainardi) [inline zkml/src/layers/transformer/mod.rs:245]: Leftover? Now causal mask is inside MHA layer, no?

File: zkml/src/layers/transformer/logits.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        match self {
+            Logits::Argmax => {
+                let indices = inputs
+                    .iter()
+                    .map(|input| {
+                        ensure!(
+                            input.get_shape().len() >= 2,
+                            "Argmax is for tensors of rank >= 2"
+                        );
+                        let last_row = input.slice_on_dim(1).last().unwrap();
```
Comment (nicholas-mainardi) [inline zkml/src/layers/transformer/logits.rs:31]: Why taking only the last row? Is it because we are assuming that all tokens except the last one are input tokens or already processed tokens?

**Review by nicholas-mainardi — COMMENTED** [review]
LGTM! Mostly minor comments

File: zkml/src/layers/matrix_mul.rs [hunk 4]
```diff
@@ -188,26 +201,33 @@ impl<T> MatMul<T> {
...
+                    );
+                }
+                OperandMatrix::Input => (),
+            }
+        }
         Ok(Self {
             left_matrix,
             right_matrix,
             config,
+            bias: None,
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/matrix_mul.rs:222]: Oupsie

File: zkml/src/layers/transformer/mha.rs [hunk 5]
```diff
@@ -203,28 +271,83 @@ mod test {
...
+        let infinitized = zeroified.add(&minus_infinity);
+        assert_eq!(zeroified.get_shape(), input.get_shape());
+        assert_eq!(infinitized.get_shape(), input.get_shape());
+        infinitized
+            .slice_on_dim(0)
+            .enumerate()
+            .all(|(head_idx, head)| {
+                head.chunks(q_len).enumerate().all(|(q_idx, q)| {
+                    q.iter().enumerate().all(|(i, v)| {
+                        if i > 0 {
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/transformer/mha.rs:344]: Fixed the test - Looking back at it, it was not testing correctly.

File: zkml/src/layers/transformer/mod.rs (hunk 2 above)
Comment (nikkolasg) [reply zkml/src/layers/transformer/mod.rs:245]: Yes

File: zkml/src/layers/transformer/logits.rs (hunk 3 above)
Comment (nikkolasg) [reply zkml/src/layers/transformer/logits.rs:31]: Well, for inference, we only care about the last one. But i agree the API could take everything...

File: zkml/src/layers/transformer/softmax.rs (hunk 1 above)
Comment (nikkolasg) [reply zkml/src/layers/transformer/softmax.rs:16]: Changed the definition of dimension, much clearer and i actually fixed a buggy test with that.

File: zkml/src/layers/transformer/logits.rs (hunk 3 above)
Comment (nikkolasg) [reply zkml/src/layers/transformer/logits.rs]: Changed it to running argmax on all the last dimension slices of all the input tensors.
